{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 265,
   "id": "34e3327f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set up packages\n",
    "import pandas as pd\n",
    "%matplotlib inline\n",
    "import matplotlib.pyplot as plt\n",
    "import re\n",
    "import random\n",
    "import numpy as np\n",
    "import os\n",
    "from functools import reduce"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 266,
   "id": "119e31b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# open the initial database\n",
    "wd = \"/Users/goldriev/Google Drive/My Drive/Developmental_competence_ms/TriTyrp_ML/\"\n",
    "df = pd.read_csv(wd + 'gene_data.csv', sep=\",\", na_values=['-'], low_memory=False)\n",
    "df.columns = [c.lower() for c in df.columns]\n",
    "df = df.replace('\\*','', regex=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 267,
   "id": "903bf37a",
   "metadata": {},
   "outputs": [],
   "source": [
    "df ['difn-bfd3'] = pd.to_numeric(df['difn/tet- (cds only)']) - pd.to_numeric(df['bfd3/tet- (cds only)']) \n",
    "df ['difn-bfd6'] = pd.to_numeric(df['difn/tet- (cds only)']) - pd.to_numeric(df['bfd6/tet- (cds only)']) \n",
    "df ['difn-pf'] = pd.to_numeric(df['difn/tet- (cds only)']) - pd.to_numeric(df['pf/tet- (cds only)']) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 268,
   "id": "d27ea205",
   "metadata": {},
   "outputs": [],
   "source": [
    "#add rnaseq results\n",
    "#mono_df = pd.read_csv('/Users/goldriev/keep/store/PHD/Monomorph_wet/Clonal_selection/BHI_selection/RNA_seq/salmon/Results/', sep=\",\", na_values=['-'], low_memory=False)\n",
    "#mono_df.columns = [c.lower() for c in mono_df.columns]\n",
    "#mono_df = mono_df.replace('\\*','', regex=True)\n",
    "#mono_df = mono_df.fillna(0)\n",
    "#mono_df[\"gene id\"] = mono_df[\"rn\"]\n",
    "#df = pd.merge(df, mono_df, how=\"left\", on=[\"gene id\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 269,
   "id": "9d44df91",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Add list of sif genes\n",
    "sifs = {'gene id':(\"Tb927.2.1810\", \"Tb927.2.2720\", \"Tb927.2.4020\", \"Tb927.3.4560\", \n",
    "        \"Tb927.4.670\", \"Tb927.4.3620\", \"Tb927.4.3630\", \"Tb927.4.3640\", \n",
    "        \"Tb927.4.3650\", \"Tb927.5.3580\", \"Tb927.6.2300\", \"Tb927.6.2360\", \n",
    "        \"Tb927.7.2100\", \"Tb927.7.7160\", \"Tb927.8.2860\", \"Tb927.9.4080\", \n",
    "        \"Tb927.9.7550\", \"Tb927.9.13530\", \"Tb927.10.5930\", \"Tb927.10.5940\", \n",
    "        \"Tb927.10.5950\", \"Tb927.10.12100\", \"Tb927.10.15020\", \"Tb927.10.16120\", \n",
    "        \"Tb927.11.290\", \"Tb927.11.300\", \"Tb927.11.750\", \"Tb927.11.760\", \n",
    "        \"Tb927.11.1640\", \"Tb927.11.2250\", \"Tb927.11.3650\", \"Tb927.11.4610\", \n",
    "        \"Tb927.11.6600\", \"Tb927.11.11470\", \"Tb927.11.11480\", \"Tb927.10.12090\", \n",
    "        \"Tb927.1.1930\", \"Tb927.11.9270\", \"Tb927.6.4220\", \"Tb927.8.1530\", \"Tb927.10.1740\", \n",
    "        \"Tb927.10.1750\", \"Tb927.10.2030\", \"Tb927.10.2530\", \"Tb927.10.12110\", \"Tb927.11.1480\", \n",
    "        \"Tb927.11.6610\", \"Tb927.7.2660\", \"Tb927.4.5390\", \"Tb927.8.6930\", \"Tb927.9.6100\", \n",
    "        \"Tb927.9.6090\", \"Tb927.8.7020\", \"Tb927.11.2500\", \"Tb927.11.12850\", \"Tb927.8.2780\")}\n",
    "sifs = pd.DataFrame(sifs)\n",
    "sifs['QS_Pathway'] = \"Y\"\n",
    "df = pd.merge(df, sifs, how=\"left\", on=[\"gene id\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 270,
   "id": "f9978ec4",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Add target\n",
    "\n",
    "#Add list of sif genes\n",
    "target = {'gene id':(\"Tb927.8.1530\", \"Tb927.5.2580\", \"Tb927.2.4020\", \"Tb927.11.6600\", \"Tb927.4.3650\", \"Tb927.11.3400\")}\n",
    "target = pd.DataFrame(target)\n",
    "target['target'] = \"Y\"\n",
    "df = pd.merge(df, target, how=\"left\", on=[\"gene id\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 271,
   "id": "polar-attempt",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Add dnds results\n",
    "directory = \"~/Desktop/\"\n",
    "code = pd.read_table(directory + \"codeml.txt\", sep = \"_\", names = [\"source_id\", \"txt\", \"dn\", \"dn/ds\"])\n",
    "dnds = code['dn'].str.split(' ',expand=True)[[5,6,7,8,9,10]]\n",
    "dnds[\"source_id\"] = code[\"source_id\"]\n",
    "dnds = dnds.rename({5:'background', 6:'ev_a', 7:'ev_b', 8:'ivmt1', 9:'ovi', 10:'botat'}, axis='columns')\n",
    "dnds[[\"background\", \"ev_a\", \"ev_b\", \"ivmt1\", \"ovi\", \"botat\"]] = dnds[[\"background\", \"ev_a\", \"ev_b\", \"ivmt1\", \"ovi\", \"botat\"]].apply(pd.to_numeric, errors='coerce')\n",
    "dnds[['gene id', 'type']] = dnds['source_id'].str.split(':', n=1, expand=True)\n",
    "df = pd.merge(df, dnds, how=\"left\", on=[\"gene id\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 272,
   "id": "welcome-faculty",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Add snpeff results\n",
    "wd = \"/Volumes/matthews/Guy/Raw_data/monomorph/data/snpeff/lineage_specific\"\n",
    "eva = pd.read_table(wd + '/T.b.evansitypeA.vcf.dir/snpEff_genes.txt', sep=\" \", na_values=['-'], low_memory=False, header = None).iloc[1:][0].str.split('\\t', expand=True)\n",
    "eva.columns = eva.iloc[0]\n",
    "eva = eva[1:].add_suffix('_eva')\n",
    "eva[['gene id', 'type']] = eva['TranscriptId_eva'].str.split(':', n=1, expand=True)\n",
    "eva.drop('type', axis=1, inplace=True)\n",
    "evb = pd.read_table(wd + '/T.b.evansitypeB.vcf.dir/snpEff_genes.txt', sep=\" \", na_values=['-'], low_memory=False, header = None).iloc[1:][0].str.split('\\t', expand=True)\n",
    "evb.columns = evb.iloc[0]\n",
    "evb = evb[1:].add_suffix('_evb')\n",
    "evb[['gene id', 'type']] = evb['TranscriptId_evb'].str.split(':', n=1, expand=True)\n",
    "evb.drop('type', axis=1, inplace=True)\n",
    "evc = pd.read_table(wd + '/T.b.evansitypeC.vcf.dir/snpEff_genes.txt', sep=\" \", na_values=['-'], low_memory=False, header = None).iloc[1:][0].str.split('\\t', expand=True)\n",
    "evc.columns = evc.iloc[0]\n",
    "evc = evc[1:].add_suffix('_evc')\n",
    "evc[['gene id', 'type']] = evc['TranscriptId_evc'].str.split(':', n=1, expand=True)\n",
    "evc.drop('type', axis=1, inplace=True)\n",
    "ovi = pd.read_table(wd + '/T.b.equitypeOVI.vcf.dir/snpEff_genes.txt', sep=\" \", na_values=['-'], low_memory=False, header = None).iloc[1:][0].str.split('\\t', expand=True)\n",
    "ovi.columns = ovi.iloc[0]\n",
    "ovi = ovi[1:].add_suffix('_ovi')\n",
    "ovi[['gene id', 'type']] = ovi['TranscriptId_ovi'].str.split(':', n=1, expand=True)\n",
    "ovi.drop('type', axis=1, inplace=True)\n",
    "botat = pd.read_table(wd + '/T.b.equitypeBOTAT.vcf.dir/snpEff_genes.txt', sep=\" \", na_values=['-'], low_memory=False, header = None).iloc[1:][0].str.split('\\t', expand=True)\n",
    "botat.columns = botat.iloc[0]\n",
    "botat = botat[1:].add_suffix('_botat')\n",
    "botat[['gene id', 'type']] = botat['TranscriptId_botat'].str.split(':', n=1, expand=True)\n",
    "botat.drop('type', axis=1, inplace=True)\n",
    "pleo = pd.read_table('/Volumes/matthews/goldriev/monomorph/snpEff_genes.txt', sep=\" \", na_values=['-'], low_memory=False, header = None).iloc[1:][0].str.split('\\t', expand=True)\n",
    "pleo.columns = pleo.iloc[0]\n",
    "pleo = pleo[1:].add_suffix('_pleo')\n",
    "pleo[['gene id', 'type']] = pleo['TranscriptId_pleo'].str.split(':', n=1, expand=True)\n",
    "pleo.drop('type', axis=1, inplace=True)\n",
    "\n",
    "dfList = [df, eva, evb, evc, ovi, botat, pleo]\n",
    "df = reduce(lambda left,right: left.merge(right, how=\"left\", on='gene id'), dfList)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 273,
   "id": "73b746e9-d55c-426d-a2a0-7cacb5480e8f",
   "metadata": {},
   "outputs": [],
   "source": [
    "df[['background','ev_a', 'ev_b', 'ivmt1', 'ovi', 'botat']] = df[['background','ev_a', 'ev_b', 'ivmt1', 'ovi', 'botat']].apply(pd.to_numeric)\n",
    "filtered = df[(df['is pseudo'] != \"Yes\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 274,
   "id": "5192461a-d472-4bd4-86ca-270ba3c05b45",
   "metadata": {},
   "outputs": [],
   "source": [
    "filtered = filtered[~filtered['product description'].str.contains('variant surface')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 275,
   "id": "8e04cd87-1824-421f-82d1-87faeedbf208",
   "metadata": {},
   "outputs": [],
   "source": [
    "filtered = filtered[filtered['chromosome'].isin(['1', '2', '3', '4', '5', '6', '7', '8', '9', '10','11'])] "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 276,
   "id": "1d9b89da-06f6-4b8c-a71f-17304504e29e",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/k5/mm9d16t101v50p5b4qwh8hfh02tcwb/T/ipykernel_35800/2081220531.py:3: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  dnds_filtered[['variants_impact_HIGH_eva', 'variants_impact_HIGH_evb', 'variants_impact_HIGH_evc', 'variants_impact_HIGH_ovi', 'variants_impact_HIGH_botat']] = dnds_filtered[['variants_impact_HIGH_eva', 'variants_impact_HIGH_evb', 'variants_impact_HIGH_evc', 'variants_impact_HIGH_ovi', 'variants_impact_HIGH_botat']].apply(pd.to_numeric)\n"
     ]
    }
   ],
   "source": [
    "filtered[['ev_a', 'variants_impact_HIGH_pleo', 'variants_impact_MODERATE_eva', 'variants_impact_MODERATE_evb', 'variants_impact_MODERATE_evc', 'variants_impact_MODERATE_ovi', 'variants_impact_MODERATE_botat']] = filtered[['ev_a', 'variants_impact_HIGH_pleo', 'variants_impact_MODERATE_eva', 'variants_impact_MODERATE_evb', 'variants_impact_MODERATE_evc', 'variants_impact_MODERATE_ovi', 'variants_impact_MODERATE_botat']].apply(pd.to_numeric)\n",
    "dnds_filtered = filtered[(filtered['background'] < 1) & (filtered['variants_impact_HIGH_pleo'] == 0 )]\n",
    "dnds_filtered[['variants_impact_HIGH_eva', 'variants_impact_HIGH_evb', 'variants_impact_HIGH_evc', 'variants_impact_HIGH_ovi', 'variants_impact_HIGH_botat']] = dnds_filtered[['variants_impact_HIGH_eva', 'variants_impact_HIGH_evb', 'variants_impact_HIGH_evc', 'variants_impact_HIGH_ovi', 'variants_impact_HIGH_botat']].apply(pd.to_numeric)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 277,
   "id": "62a375ac-e5bc-4640-945e-24ded22ed1a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "dnds_filtered = dnds_filtered[(dnds_filtered['ev_a'] > 1) | (dnds_filtered['ev_b'] > 1) | (dnds_filtered['ivmt1'] > 1) | (dnds_filtered['ovi'] > 1) | (dnds_filtered['botat'] > 1)]\n",
    "dnds_filtered = dnds_filtered[(dnds_filtered['variants_impact_MODERATE_eva'] > 0) | (dnds_filtered['variants_impact_MODERATE_evb'] > 0) | (dnds_filtered['variants_impact_MODERATE_evc'] > 0) | (dnds_filtered['variants_impact_MODERATE_ovi'] > 0) | (dnds_filtered['variants_impact_MODERATE_botat'] > 0) | (dnds_filtered['variants_impact_HIGH_eva'] > 0) | (dnds_filtered['variants_impact_HIGH_evb'] > 0) | (dnds_filtered['variants_impact_HIGH_evc'] > 0) | (dnds_filtered['variants_impact_HIGH_ovi'] > 0) | (dnds_filtered['variants_impact_HIGH_botat'] > 0)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 278,
   "id": "a26598e8-7001-42b4-b560-9db5614ecf88",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene id</th>\n",
       "      <th># exons in gene</th>\n",
       "      <th># exons in transcript</th>\n",
       "      <th># transcripts</th>\n",
       "      <th>annotated 3' utr length</th>\n",
       "      <th>annotated 5' utr length</th>\n",
       "      <th>gene strand</th>\n",
       "      <th>gene type</th>\n",
       "      <th>is pseudo</th>\n",
       "      <th>transcript length</th>\n",
       "      <th>...</th>\n",
       "      <th>variants_effect_splice_acceptor_variant_pleo</th>\n",
       "      <th>variants_effect_splice_donor_variant_pleo</th>\n",
       "      <th>variants_effect_splice_region_variant_pleo</th>\n",
       "      <th>variants_effect_start_lost_pleo</th>\n",
       "      <th>variants_effect_start_retained_variant_pleo</th>\n",
       "      <th>variants_effect_stop_gained_pleo</th>\n",
       "      <th>variants_effect_stop_lost_pleo</th>\n",
       "      <th>variants_effect_stop_retained_variant_pleo</th>\n",
       "      <th>variants_effect_synonymous_variant_pleo</th>\n",
       "      <th>variants_effect_upstream_gene_variant_pleo</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1859</th>\n",
       "      <td>Tb927.1.1060</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>832.0</td>\n",
       "      <td>67.0</td>\n",
       "      <td>forward</td>\n",
       "      <td>protein coding gene</td>\n",
       "      <td>No</td>\n",
       "      <td>1817</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1880</th>\n",
       "      <td>Tb927.1.1330</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>19.0</td>\n",
       "      <td>forward</td>\n",
       "      <td>protein coding gene</td>\n",
       "      <td>No</td>\n",
       "      <td>1240</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1910</th>\n",
       "      <td>Tb927.1.1750</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>16.0</td>\n",
       "      <td>44.0</td>\n",
       "      <td>forward</td>\n",
       "      <td>protein coding gene</td>\n",
       "      <td>No</td>\n",
       "      <td>1473</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1915</th>\n",
       "      <td>Tb927.1.1820</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>42.0</td>\n",
       "      <td>152.0</td>\n",
       "      <td>forward</td>\n",
       "      <td>protein coding gene</td>\n",
       "      <td>No</td>\n",
       "      <td>1340</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>129</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1964</th>\n",
       "      <td>Tb927.1.2570</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>23.0</td>\n",
       "      <td>forward</td>\n",
       "      <td>protein coding gene</td>\n",
       "      <td>No</td>\n",
       "      <td>2972</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11479</th>\n",
       "      <td>Tb927.9.8450</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>275.0</td>\n",
       "      <td>92.0</td>\n",
       "      <td>reverse</td>\n",
       "      <td>protein coding gene</td>\n",
       "      <td>No</td>\n",
       "      <td>970</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11495</th>\n",
       "      <td>Tb927.9.8680</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>51.0</td>\n",
       "      <td>reverse</td>\n",
       "      <td>protein coding gene</td>\n",
       "      <td>No</td>\n",
       "      <td>867</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>79</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11555</th>\n",
       "      <td>Tb927.9.9550</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>388.0</td>\n",
       "      <td>112.0</td>\n",
       "      <td>reverse</td>\n",
       "      <td>protein coding gene</td>\n",
       "      <td>No</td>\n",
       "      <td>1139</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>62</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11563</th>\n",
       "      <td>Tb927.9.9660</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>926.0</td>\n",
       "      <td>81.0</td>\n",
       "      <td>reverse</td>\n",
       "      <td>protein coding gene</td>\n",
       "      <td>No</td>\n",
       "      <td>2063</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11564</th>\n",
       "      <td>Tb927.9.9670</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>372.0</td>\n",
       "      <td>170.0</td>\n",
       "      <td>reverse</td>\n",
       "      <td>protein coding gene</td>\n",
       "      <td>No</td>\n",
       "      <td>1295</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>540 rows Ã— 257 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            gene id  # exons in gene  # exons in transcript  # transcripts  \\\n",
       "1859   Tb927.1.1060                1                      1              1   \n",
       "1880   Tb927.1.1330                1                      1              1   \n",
       "1910   Tb927.1.1750                1                      1              1   \n",
       "1915   Tb927.1.1820                1                      1              1   \n",
       "1964   Tb927.1.2570                1                      1              1   \n",
       "...             ...              ...                    ...            ...   \n",
       "11479  Tb927.9.8450                1                      1              1   \n",
       "11495  Tb927.9.8680                1                      1              1   \n",
       "11555  Tb927.9.9550                1                      1              1   \n",
       "11563  Tb927.9.9660                1                      1              1   \n",
       "11564  Tb927.9.9670                1                      1              1   \n",
       "\n",
       "       annotated 3' utr length  annotated 5' utr length gene strand  \\\n",
       "1859                     832.0                     67.0     forward   \n",
       "1880                       NaN                     19.0     forward   \n",
       "1910                      16.0                     44.0     forward   \n",
       "1915                      42.0                    152.0     forward   \n",
       "1964                       NaN                     23.0     forward   \n",
       "...                        ...                      ...         ...   \n",
       "11479                    275.0                     92.0     reverse   \n",
       "11495                      NaN                     51.0     reverse   \n",
       "11555                    388.0                    112.0     reverse   \n",
       "11563                    926.0                     81.0     reverse   \n",
       "11564                    372.0                    170.0     reverse   \n",
       "\n",
       "                 gene type is pseudo  transcript length  ...  \\\n",
       "1859   protein coding gene        No               1817  ...   \n",
       "1880   protein coding gene        No               1240  ...   \n",
       "1910   protein coding gene        No               1473  ...   \n",
       "1915   protein coding gene        No               1340  ...   \n",
       "1964   protein coding gene        No               2972  ...   \n",
       "...                    ...       ...                ...  ...   \n",
       "11479  protein coding gene        No                970  ...   \n",
       "11495  protein coding gene        No                867  ...   \n",
       "11555  protein coding gene        No               1139  ...   \n",
       "11563  protein coding gene        No               2063  ...   \n",
       "11564  protein coding gene        No               1295  ...   \n",
       "\n",
       "      variants_effect_splice_acceptor_variant_pleo  \\\n",
       "1859                                             0   \n",
       "1880                                             0   \n",
       "1910                                             0   \n",
       "1915                                             0   \n",
       "1964                                             0   \n",
       "...                                            ...   \n",
       "11479                                            0   \n",
       "11495                                            0   \n",
       "11555                                            0   \n",
       "11563                                            0   \n",
       "11564                                            0   \n",
       "\n",
       "      variants_effect_splice_donor_variant_pleo  \\\n",
       "1859                                          0   \n",
       "1880                                          0   \n",
       "1910                                          0   \n",
       "1915                                          0   \n",
       "1964                                          0   \n",
       "...                                         ...   \n",
       "11479                                         0   \n",
       "11495                                         0   \n",
       "11555                                         0   \n",
       "11563                                         0   \n",
       "11564                                         0   \n",
       "\n",
       "      variants_effect_splice_region_variant_pleo  \\\n",
       "1859                                           0   \n",
       "1880                                           0   \n",
       "1910                                           0   \n",
       "1915                                           0   \n",
       "1964                                           0   \n",
       "...                                          ...   \n",
       "11479                                          0   \n",
       "11495                                          0   \n",
       "11555                                          0   \n",
       "11563                                          0   \n",
       "11564                                          0   \n",
       "\n",
       "      variants_effect_start_lost_pleo  \\\n",
       "1859                                0   \n",
       "1880                                0   \n",
       "1910                                0   \n",
       "1915                                0   \n",
       "1964                                0   \n",
       "...                               ...   \n",
       "11479                               0   \n",
       "11495                               0   \n",
       "11555                               0   \n",
       "11563                               0   \n",
       "11564                               0   \n",
       "\n",
       "      variants_effect_start_retained_variant_pleo  \\\n",
       "1859                                            0   \n",
       "1880                                            0   \n",
       "1910                                            0   \n",
       "1915                                            0   \n",
       "1964                                            0   \n",
       "...                                           ...   \n",
       "11479                                           0   \n",
       "11495                                           0   \n",
       "11555                                           0   \n",
       "11563                                           0   \n",
       "11564                                           0   \n",
       "\n",
       "      variants_effect_stop_gained_pleo variants_effect_stop_lost_pleo  \\\n",
       "1859                                 0                              0   \n",
       "1880                                 0                              0   \n",
       "1910                                 0                              0   \n",
       "1915                                 0                              0   \n",
       "1964                                 0                              0   \n",
       "...                                ...                            ...   \n",
       "11479                                0                              0   \n",
       "11495                                0                              0   \n",
       "11555                                0                              0   \n",
       "11563                                0                              0   \n",
       "11564                                0                              0   \n",
       "\n",
       "      variants_effect_stop_retained_variant_pleo  \\\n",
       "1859                                           0   \n",
       "1880                                           0   \n",
       "1910                                           0   \n",
       "1915                                           0   \n",
       "1964                                           0   \n",
       "...                                          ...   \n",
       "11479                                          0   \n",
       "11495                                          0   \n",
       "11555                                          0   \n",
       "11563                                          0   \n",
       "11564                                          0   \n",
       "\n",
       "      variants_effect_synonymous_variant_pleo  \\\n",
       "1859                                        1   \n",
       "1880                                        0   \n",
       "1910                                        6   \n",
       "1915                                        5   \n",
       "1964                                        5   \n",
       "...                                       ...   \n",
       "11479                                       3   \n",
       "11495                                       3   \n",
       "11555                                       3   \n",
       "11563                                       3   \n",
       "11564                                       2   \n",
       "\n",
       "       variants_effect_upstream_gene_variant_pleo  \n",
       "1859                                            4  \n",
       "1880                                           21  \n",
       "1910                                           25  \n",
       "1915                                          129  \n",
       "1964                                            8  \n",
       "...                                           ...  \n",
       "11479                                          49  \n",
       "11495                                          79  \n",
       "11555                                          62  \n",
       "11563                                          29  \n",
       "11564                                          13  \n",
       "\n",
       "[540 rows x 257 columns]"
      ]
     },
     "execution_count": 278,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dnds_filtered = dnds_filtered[(dnds_filtered['difn-bfd3'] <= 0) & (dnds_filtered['difn-bfd6'] <= 0) & (dnds_filtered['difn-pf'] <= 0)]\n",
    "dnds_filtered[['difn/tet- (cds only)']] = dnds_filtered[['difn/tet- (cds only)']].apply(pd.to_numeric)\n",
    "dnds_filtered = dnds_filtered[(dnds_filtered['difn/tet- (cds only)'] < -1.5)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 264,
   "id": "6e569607-4d2f-43cd-8432-531351e480d0",
   "metadata": {},
   "outputs": [],
   "source": [
    "dnds_filtered.to_csv('~/Desktop/dnds_targets.csv', index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 290,
   "id": "477e65cd-8b75-4030-9402-35ab7f239549",
   "metadata": {},
   "outputs": [],
   "source": [
    "qs_filtered = filtered\n",
    "qs_filtered[['ev_a', 'variants_impact_HIGH_pleo', 'variants_impact_MODERATE_eva', 'variants_impact_MODERATE_evb', 'variants_impact_MODERATE_evc', 'variants_impact_MODERATE_ovi', 'variants_impact_MODERATE_botat']] = qs_filtered[['ev_a', 'variants_impact_HIGH_pleo', 'variants_impact_MODERATE_eva', 'variants_impact_MODERATE_evb', 'variants_impact_MODERATE_evc', 'variants_impact_MODERATE_ovi', 'variants_impact_MODERATE_botat']].apply(pd.to_numeric)\n",
    "qs_filtered[['variants_impact_HIGH_eva', 'variants_impact_HIGH_evb', 'variants_impact_HIGH_evc', 'variants_impact_HIGH_ovi', 'variants_impact_HIGH_botat']] = qs_filtered[['variants_impact_HIGH_eva', 'variants_impact_HIGH_evb', 'variants_impact_HIGH_evc', 'variants_impact_HIGH_ovi', 'variants_impact_HIGH_botat']].apply(pd.to_numeric)\n",
    "qs_filtered = qs_filtered[(qs_filtered['variants_impact_HIGH_pleo'] == 0 )]\n",
    "qs_filtered = filtered[(filtered['variants_impact_MODERATE_eva'] > 0) | (filtered['variants_impact_MODERATE_evb'] > 0) | (filtered['variants_impact_MODERATE_evc'] > 0) | (filtered['variants_impact_MODERATE_ovi'] > 0) | (filtered['variants_impact_MODERATE_botat'] > 0) | (filtered['variants_impact_HIGH_eva'] > 0) | (dnds_filtered['variants_impact_HIGH_evb'] > 0) | (dnds_filtered['variants_impact_HIGH_evc'] > 0) | (dnds_filtered['variants_impact_HIGH_ovi'] > 0) | (filtered['variants_impact_HIGH_botat'] > 0)]\n",
    "qs_filtered = qs_filtered[(qs_filtered['QS_Pathway'] == \"Y\")]\n",
    "qs_filtered.to_csv('~/Desktop/qs_targets.csv', index = False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
